cs3api4lab/api/cs3_share_api.py

```diff
@@ -141,36 +146,101 @@ def update(self, share_id, role):
+            storage_id=storage_id,
+            opaque_id=opaque_id
+        )
+        share_filters.append(sharing_res.Filter(
+            resource_id=resource,
+            type=sharing_res.Filter.Type.TYPE_RESOURCE_ID
+        ))
+
+        list_response = self.cs3_api.ListReceivedShares(
+            request=sharing.ListReceivedSharesRequest(filters=share_filters),
+            metadata=[('x-access-token', self.auth.authenticate())]
+        )
+        share = None
+        if len(list_response.shares):
+            share = list_response.shares.pop().share
+        return share  # this should return only one share
+
+    def list_received(self, path=None):
         self.log.info("Listing received shares")
-        list_request = sharing.ListReceivedSharesRequest()
+        list_request = self._shares_received_filter_by_resource(path)
         list_response = self.cs3_api.ListReceivedShares(request=list_request,
                                                         metadata=[('x-access-token', self.auth.authenticate())])
 
-        if self._is_code_ok(list_response):
-            self.log.debug(f"Retrieved received shares for user {self.config.client_id}:\n{list_response}")
-            return list_response
-        else:
+        if not self._is_code_ok(list_response):
             self.log.error("Error retrieving received shares for user: " + self.config.client_id)
             self._handle_error(list_response)
 
+        self.log.debug(f"Retrieved received shares for user {self.config.client_id}:\n{list_response}")
+        return list_response
+
     def _share_filter_by_resource(self, path):
         if path is None:
             return sharing.ListSharesRequest()
 
         file_stat = self.file_api.stat_info(path)
-        share_filters = []
+        opaque_id = urllib.parse.unquote(file_stat['inode']['opaque_id'])
+        storage_id = urllib.parse.unquote(file_stat['inode']['storage_id'])
+
+        # remove after https://github.com/cs3org/reva/issues/3243 is fixed
+        opaque_id = FileUtils.fix_dev_opaque(opaque_id, self.config.dev_env)
+
         resource = storage_resources.ResourceId(
-            storage_id=file_stat['inode']['storage_id'],
-            opaque_id=file_stat['inode']['opaque_id']
+            storage_id=storage_id,
+            opaque_id=opaque_id
         )
+
+        share_filters = []
         share_filters.append(sharing_res.Filter(
             resource_id=resource,
             type=sharing_res.Filter.Type.TYPE_RESOURCE_ID
         ))
-        list_request = sharing.ListSharesRequest(filters=share_filters)
 
-        return list_request
+        return sharing.ListSharesRequest(filters=share_filters)
+
+    def _shares_received_filter_by_resource(self, path):
+        if path is None:
+            return sharing.ListReceivedSharesRequest()
+
+        try:
+            file_stat = self.file_api.stat_info(path)
+        except FileNotFoundError:
+            return sharing.ListSharesRequest()
```

Comment:
The stat request is here because we use paths that are problematic.

Depending on REVA configuration paths in REVA are interpreted differently. For example in two-server-config REVA configuration example resources for shares and files on the user account have different paths /home and /reva. At this point, Reva thinks that the same shared file accessed from /home and from /reva path are different resources. 

This is why we are using opaque_id and storage_id. This is the safest way to get information about resources.

This filter is used to determine if the file we're trying to open/save is shared and check if we have permission to save a file or open it in read-only mode.

Reply:
Please document it in the code with the full explanation. If I understood correctly, is that the filtering needs to be done by id and not by path, because the same id can have multiple paths, correct?

Reply:
It's because it's used to build filters for list_share. Share resource is defined by three parameters (owner, resource_id, grantees) which we don't have in this case so I could not use the GetShare endpoint. I needed a way to get a single share by a path (this was created instead of listing all of the shares and filtering a shared later). I believe this was the only way to get information about share that I needed.

Reply:
@diocas  " If I understood correctly, is that the filtering needs to be done by id and not by path, because the same id can have multiple paths, correct?"
In two-server-config for shares to work we use local and localhost drivers. Local is used for handling user home dir and localhost is used to handle shares. So if I share a resource and try to open it from /home it has a different ID than when i try to open a share from shares panel /reva/username (localhome driver). So depending on the path we use to open a resource (even tho it's the same file) REVA does not recognize it's the same file (for example metadata are not visible). So until the bug is fixed the only safe way is to use the ID (which I'm fixing in code)